### PR TITLE
Fix broken parallelism in quic-client

### PR DESCRIPTION
**Pull request by ripatel-fd**
Fixes excessive fragmentation by TPU clients leading to a large
number of streams per conn in 'sending' state simultaneously.
This, in turn, requires excessive in-memory buffering server-side
to reassemble fragmented transactions.

- Simplifies QuicClient::send_batch to enqueue send operations
  in sequential order
- Removes the "max_parallel_streams" config option

The quic-client now produces an ordered fragment stream when
scheduling send operations from a single-thread.

### CODE REVIEW

File: quic-client/src/nonblocking/quic_client.rs
Comment (lijunwangs) [inline quic-client/src/nonblocking/quic_client.rs:580]: Why is max_parallel_streams removed?

Comment (ripatel-fd) [reply quic-client/src/nonblocking/quic_client.rs:580]: @lijunwangs The logic that read this variable was removed entirely so the setting no longer does anything. 
The QUIC client now sends each stream sequentially. This is faster and simpler than the fragmentation that the parallel setting created. The trick to make this work is that the client no longer waits for acknowledgements.

**Comment by ripatel-fd** [comment]
Let's hold off on merging this until we've gathered some benchmarks. I suggest
1. bench-tps against a local test validator (to measure max throughput under optimal conditions) 
2. bench-tps against a remote validator transatlantic (to see if this change is sensitive to latency)
3. bench-tps against a Firedancer v0.1 validator. The Firedancer team can help test this against a cluster instead of just individual validators for more realistic results
4. Blind spam against fd_quic (I already wrote this test)

For each test, I'll attach bench-tps reported counters and fragmentation rates (depends on #2539).

@lijunwangs @alessandrod Can you think of any other tests or datapoints we'd need?

File: quic-client/src/nonblocking/quic_client.rs
Comment (ripatel-fd) [inline quic-client/src/nonblocking/quic_client.rs]: We should probably report the bug that requires this priority hack to the quinn maintainers.
There shouldn't be any reordering in a simple loop like this:

```rust
loop {
    let mut send_stream = connection.open_uni().await?;
    send_stream.write_all(some_buffer()).await?;
}
```

Comment (alessandrod) [reply quic-client/src/nonblocking/quic_client.rs]: why is this needed?

**Review by alessandrod — COMMENTED** [review]
This looks good, just wondering why the set_priority thing is needed

**Comment by ripatel-fd** [comment]
> why is [set_priority] needed?

@alessandrod It's a workaround for a quinn bug. 

If you do this you get reordering much worse than the parallel send method before this patch.

```rust
loop {
    let mut send_stream = connection.open_uni().await?;
    send_stream.write_all(data).await?;
}
```

I suspect the reason is that the `WriteAll` future completes when some stream data was not serialized to QUIC frames and some internal state object for that stream sticks around. As the loop iterations run, those internal stream objects then stick around in some unordered data structure. I added the `set_priority` statement as an experiment and it fixes the issue. 

Here is a writeup on the issue: [Firedancer-2024-08-10 solana-quic-client fragmentation-110824-162447.pdf](https://github.com/user-attachments/files/16843457/Firedancer-2024-08-10.solana-quic-client.fragmentation-110824-162447.pdf)

**Review by alessandrod — APPROVED** [review]
I have rebased this and removed set_priority since I'm not seeing the issue @ripatel-fd was seeing on fd. The PR is obviously correct. I've been doing all my testing in the last month with similar code (no parallel streams).

**Review by alessandrod — CHANGES_REQUESTED** [review]
Tests need work, I'll fixup

**Review by alessandrod — APPROVED** [review]
ok good to go now

**Comment by lijunwangs** [comment]
Can you show some of the results of testing with/without the changes?

**Comment by KirillLykov** [comment]
I'm using similar code for quite some time, see https://github.com/anza-xyz/agave/pull/2905 and it respects flow control properly (while multistream doesn't not).

> 1. bench-tps against a local test validator (to measure max throughput under optimal conditions)

In this case the connection will be limited by `512` concurrent streams and `512*1232` receive window size. So this setup will not show max throughput. What can be done instead is using some mock server which respects the protocol but doesn't limit the connection.
 
> 2. bench-tps against a remote validator transatlantic (to see if this change is sensitive to latency).

Well, what is sensitive to latency is our server side because we fix RW to `512*1232` regardless of rtt.

**Comment by KirillLykov** [comment]
I'm confused now, what is the motivation for `quic-client` crate? It looks like it is not used in tpu-client. Instead, there is approximately the same code as was in quic-client:
https://github.com/anza-xyz/agave/blob/489f483e1d7b30ef114e0123994818b2accfa389/tpu-client/src/nonblocking/tpu_client.rs#L452

**Comment by KirillLykov** [comment]
#### Setup

Server is in Korea, client in Amsterdam. RTT: 251ms

#### Run 1: new vs old with staked connection (100% stake)

The connection is limited by our qos: `512` streams, `512*1232` RW size.
As expected, there is no difference between two runs: in both cases tps is ~16k.

On the plot below the new code is on the left, the old on the right:

Screenshot 2024-10-02 at 11 24 20

#### Run 2: new vs old with mock-server

In parallel with agave-validator on the same machine I launched mock-server which listens port 1234. 
On the client side, I modified TPU client to send txs to 1234 instead of port obtained from gossip so that bench-tps uses normal rpc but our mock server as TPU endpoint.

In this case, TPS with old client is `149520`, while with new is `165520` (10% improvement).

#### How I was running it

Client arguments:
```shell
-u http://<IP>:8899 --identity /home/klykov/wrks_ext/bench-tps-dos-test/test-for-pr2526/testnet-dos-funder.json --read-client-keys accounts.yml --duration 320 --tx_count 5000 --keypair-multiplier 2 --thread-batch-sleep-ms 10 --client-node-id id.json --block-data-file block.csv --use-tpu-client --threads 4 --tpu-connection-pool-size 4 --sustained
```

Agave server arguments (configured with multinode scripts):
```shell
./target/release/agave-validator --enable-rpc-transaction-history --enable-extended-tx-metadata-storage --require-tower --ledger /home/klykov/sol/wrks_ext/solana/net/../config/bootstrap-validator --rpc-port 8899 --snapshot-interval-slots 200 --no-incremental-snapshots --identity /home/klykov/sol/wrks_ext/solana/net/../config/bootstrap-validator/identity.json --vote-account /home/klykov/sol/wrks_ext/solana/net/../config/bootstrap-validator/vote-account.json --rpc-faucet-address 127.0.0.1:9900 --no-poh-speed-test --no-os-network-limits-test --no-wait-for-vote-to-start-leader --full-rpc-api --allow-private-addr --gossip-port 8001 --gossip-host <PUBLIC_IPv4>  --log - 2> err.txt
```

Mock server arguments:
```shell
./target/release/server --listen 0.0.0.0:1234 --receive-window-size 630784000  --max-concurrent-streams 512000 --stream-receive-window-size 1232000
```

**Review by lijunwangs — APPROVED** [review]

**Pull request merged**
